ui/src/__tests__/spec/wdio.conf.js

```diff
@@ -84,158 +156,322 @@ const config = {
+     * for that worker as well as modify runtime environments in an async fashion.
+     * @param  {String} cid      capability id (e.g 0-0)
+     * @param  {[type]} caps     object containing capabilities for session that will be spawn in the worker
+     * @param  {[type]} specs    specs to be run in the worker process
+     * @param  {[type]} args     object that will be merged with the main configuration once worker is initialized
+     * @param  {[type]} execArgv list of string arguments passed to the worker process
+     */
+    // onWorkerStart: function (cid, caps, specs, args, execArgv) {
+    // },
+    /**
+     * Gets executed just after a worker process has exited.
+     * @param  {String} cid      capability id (e.g 0-0)
+     * @param  {Number} exitCode 0 - success, 1 - fail
+     * @param  {[type]} specs    specs to be run in the worker process
+     * @param  {Number} retries  number of retries used
+     */
+    // onWorkerEnd: function (cid, exitCode, specs, retries) {
     // },
     /**
      * Gets executed just before initialising the webdriver session and test framework. It allows you
      * to manipulate configurations depending on the capability or spec.
      * @param {Object} config wdio configuration object
      * @param {Array.<Object>} capabilities list of capabilities details
      * @param {Array.<String>} specs List of spec file paths that are to be run
+     * @param {String} cid worker id (e.g. 0-0)
      */
-    // beforeSession: function (config, capabilities, specs) {
+    // beforeSession: function (config, capabilities, specs, cid) {
     // },
     /**
      * Gets executed before test execution begins. At this point you can access to all global
      * variables like `browser`. It is the perfect place to define custom commands.
      * @param {Array.<Object>} capabilities list of capabilities details
-     * @param {Array.<String>} specs List of spec file paths that are to be run
+     * @param {Array.<String>} specs        List of spec file paths that are to be run
+     * @param {Object}         browser      instance of created browser/device session
      */
-    // before: function(capabilities, specs) {
+     before: function(capabilities, specs, browser) {
+        browser.addCommand('screenshot', function(filename) {
+          return browser.takeScreenshot().then(function(png) {
+            return browser.driver.controlFlow().execute(function() {
+              return new Promise(function(fulfill) {
+                mkdirp(path.dirname(filename), function() {
+                  let stream = fs.createWriteStream(filename);
+                  stream.write(new Buffer(png, 'base64'));
+                  stream.end(function() {
+                    fulfill();
+                  });
+                });
+              });
+            });
+          });
+        });
+        let count = 0;
+        let Mocha = require('mocha');
+        if (!Mocha._pesExt) {
```

Comment:
what is _pesExt ?

Reply:
it's just a flag to check if the hook is applied onto mocha. I actually don't think it's needed since saucelabs records and saves all test failures which means this screenshot logic is not needed. I'll remove this